publisher/publisher.go

```diff
@@ -0,0 +1,288 @@
+				// Retry the request after either 10 seconds or the period specified
+				// by the Retry-After header
+				backoff := pub.CT.SubmissionBackoff
+				if retryAfter := resp.Header.Get("Retry-After"); retryAfter != "" {
+					if seconds, err := strconv.Atoi(retryAfter); err != nil {
+						backoff = time.Second * time.Duration(seconds)
+					}
+				}
+				if retries >= pub.CT.SubmissionRetries {
+					break
+				}
+				retries++
+				time.Sleep(backoff)
+				continue
+			} else if resp.StatusCode != http.StatusOK {
+				// Not something we expect to happen, set error, break loop and log
+				// the error
+				// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+				pub.log.AuditErr(fmt.Errorf("Unexpected status code returned from CT log submission endpoint [%s]: Unexpected status code [%d]", log.URI, resp.StatusCode))
+				break
+			}
+		}
+
+		done = true
+		break
+	}
+	if !done {
+		pub.log.Warning(fmt.Sprintf(
+			"Unable to submit certificate to CT log [Serial: %s, Log URI: %s, Retries: %d]",
+			serial,
+			log.URI,
+			retries,
+		))
+		return fmt.Errorf("Unable to submit certificate")
+	}
+
+	if err := sct.CheckSignature(); err != nil {
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		pub.log.AuditErr(err)
+		return err
+	}
+
+	pub.log.Notice(fmt.Sprintf(
+		"Submitted certificate to CT log [Serial: %s, Log URI: %s, Retries: %d, Signature: %x]",
+		serial,
+		log.URI,
+		retries, sct.Signature,
+	))
+
+	// Set certificate serial and add SCT to DB
+	sct.CertificateSerial = serial
+	err := pub.SA.AddSCTReceipt(sct)
+	if err != nil {
+		if strings.Contains(strings.ToLower(err.Error()), "duplicate entry") {
+			// We might want to check that the returned SCT == the SCT that we already
+			// have on file, but for now just ignore.
+			pub.log.Notice(fmt.Sprintf(
+				"SCT receipt has previously been submitted & stored [Serial: %s, Log URI: %s]",
+				serial,
+				log.URI,
+			))
+			return nil
+		}
+		// AUDIT[ Error Conditions ] 9cc4d537-8534-4970-8665-4b382abe82f3
+		pub.log.AuditErr(fmt.Errorf(
+			"Error adding SCT receipt for [%s to %s]: %s",
+			sct.CertificateSerial,
+			log.URI,
+			err,
+		))
+		return err
+	}
+	pub.log.Notice(fmt.Sprintf(
+		"Stored SCT receipt from CT log submission [Serial: %s, Log URI: %s]",
+		serial,
+		log.URI,
+	))
+	return nil
+}
+
+// SubmitToCT will submit the certificate represented by certDER to any CT
+// logs configured in pub.CT.Logs
+func (pub *PublisherAuthorityImpl) SubmitToCT(der []byte) error {
+	if pub.CT == nil {
+		return nil
+	}
+
+	cert, err := x509.ParseCertificate(der)
+	if err != nil {
+		pub.log.Err(fmt.Sprintf("Unable to parse certificate, %s", err))
+		return err
+	}
+
+	submission := ctSubmissionRequest{Chain: []string{base64.StdEncoding.EncodeToString(cert.Raw)}}
+	// Add all intermediate certificates needed for submission
+	submission.Chain = append(submission.Chain, pub.CT.IssuerBundle...)
+	client := http.Client{}
```

Comment:
You should create this on intialization and store it on the publisher object. HTTP clients are safer for use by multiple goroutines. https://golang.org/pkg/net/http/#Client